### PR TITLE
Style map

### LINKED ISSUE #12: Add style maps
Add style map functionality:
* https://github.com/allenai/hierplane#style-maps

### PR DESCRIPTION
**Pull request by mathidachuk**
Add interface for passing styling options to hierplane function. Add styling options list as guidance for user. 

see `wip.r` for list of usage examples

TODO: test + documentation. 

Wanted to get your feedback on implementation before adding tests and documentation.

Closes #12

### CODE REVIEW

File: wip.R [hunk 1]
```diff
@@ -0,0 +1,103 @@
+invisible(lapply(list.files("R", full.names = T), source))
```
[the remark this answers is not in the document]
Comment (tyluRp) [reply wip.R]: If you want, you can copy/paste the WIP code in the PR too!

**Comment by mathidachuk** [comment]
saving for laterrrr
```
invisible(lapply(list.files("R", full.names = T), source))

a <- as.data.frame(dplyr::tribble(
  ~parent_id,  ~child_id,    ~child,      ~node_type,     ~link,      ~height,  ~age,
  "Bob",       "Bob",        "Bob",       "gen1",         "ROOT",     "100 cm", "60 yo",
  "Bob",       "Angelica",   "Angelica",  "gen2",         "daughter", "100 cm", "30 yo",
  "Bob",       "Eliza",      "Eliza",     "gen2",         "daughter", "90 cm",  "25 yo",
  "Bob",       "Peggy",      "Peggy",     "gen2",         "daughter", "50 cm",  "10 yo",
  "Angelica",  "John",       "John",      "gen3",         "son",      "10 cm",  "0.5 yo",
  "Bob",       "Jess",       "Jess",      "ext",          "cousin",   "70 cm",  "150 yo"
))

hierplane(a, title = "Family Tree",
          settings = hierplane_settings(attributes = c("height", "age")))
hierplane_spacy("Bob has three daughters and a grandson.")

# ----
# styles from list
library(magrittr) # for pipe
node_type_to_style <- list("gen1" = c("color1", "strong"),
                           "gen2" = list("color6"),
                           "gen3" = list("color4"),
                           "ext" = list("placeholder"))
link_to_positions <- list(cousin = "right")
link_name_to_label <- list(daughter = "dtr", son = "son", cousin = "cuz")
settings <- hierplane_settings(attributes = c("height", "age")) %>%
  add_styles(
    node_type_to_style = node_type_to_style,
    link_to_positions = link_to_positions,
    link_name_to_label = link_name_to_label
  )

hierplane(a, title = "Family Tree", settings = settings, theme = "dark")

# check styling input and show warnings for unmatched values
# hierplane will still plot, but unmatched are ignored hence the warnings
# TLDR bad LHS assignments throw warnings
node_type_to_style <- list("gen1" = c("color1", "strong", "color2"),
                           "gen2" = list("color6"),
                           "gen2" = list("color4"), # repeat label
                           "ext" = list("placeholder"))
link_to_positions <- list(cousssssin = "right") # bad name
link_name_to_label <- list(daughter = "D", son = "S", son = "SS") # repeat label
settings <- hierplane_settings(attributes = c("height", "age")) %>%
  add_styles(
    node_type_to_style = node_type_to_style,
    link_to_positions = link_to_positions,
    link_name_to_label = link_name_to_label
  )

hierplane(a, title = "Family Tree", settings = settings, theme = "dark")

# check styling input and throw error mismatched styling options
# hierplane will fail to plot hence the errors
# TLDR bad RHS assignments throw errors

## example 1
node_type_to_style <- list("gen1" = c("color1", "strong", "color2"),
                           "gen2" = list("colr6"), # bad option
                           "ext" = list("placeholder"))
settings <- hierplane_settings(attributes = c("height", "age")) %>%
  add_styles(
    node_type_to_style = node_type_to_style
  )

hierplane(a, title = "Family Tree", settings = settings, theme = "dark")

## example 2
link_to_positions <- list(cousssssin = "top") # bad option
settings <- hierplane_settings(attributes = c("height", "age")) %>%
  add_styles(
    link_to_positions = link_to_positions
  )

hierplane(a, title = "Family Tree", settings = settings, theme = "dark")

# spacy styling demo
node_type_to_style <- list(ROOT = list("color1", "strong"),
                           dobj = list("color6"),
                           nsubj = list("color2"),
                           punct = list("placeholder"))
link_to_positions <- list(nsubj = "left", dobj = "right")
link_name_to_label <- list(nsubj = "subj", dobj = "obj")
settings <- spacy_default() %>%
  add_styles(node_type_to_style = node_type_to_style,
             link_to_positions = link_to_positions,
             link_name_to_label = link_name_to_label)

hierplane_spacy("Bob has three daughters and a grandson.",
                settings = settings, theme = "dark")

# bad nesting that is fixed automatically in format_style
node_type_to_style <- list(ROOT = c("color1", "strong"),
                           dobj = list("color6"),
                           nsubj = "color2",
                           punct = "placeholder")
settings <- spacy_default() %>%
  add_styles(node_type_to_style = node_type_to_style,
             link_to_positions = link_to_positions,
             link_name_to_label = link_name_to_label)

hierplane_spacy("Bob has three daughters and a grandson.",
                settings = settings, theme = "dark")

```

File: wip.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (mathidachuk) [reply wip.R]: added code and all the rcmd stuff *should* be addressed now

File: R/add_styles.R [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    return(x)
+  }
+
+}
+
+check_style <- function(x, settings, setting_type, setting_target) {
+
+  selected <- settings[[setting_type]]
+  bad_vals <- setdiff(names(selected), x[[settings[[setting_target]]]])
+  bad_settings <- setdiff(unlist(selected), style[[setting_type]])
```
Comment (tyluRp) [inline R/add_styles.R:47]: Should this be `x[[setting_type]]` instead of `style[[setting_type]]`?

Comment (mathidachuk) [reply R/add_styles.R:47]: This is trying to check the values selected in the settings against the valid options defined in `styles`. I think it should be `styles[[setting_type]]`. Also while typing this I realized I made a typo and it should be `styles` instead of `style`. whoops

Comment (tyluRp) [reply R/add_styles.R]: Got it, this once you change to `styles`, RCMD should pass 😸

Comment (mathidachuk) [reply R/add_styles.R]: 🤦

Comment (tyluRp) [reply R/add_styles.R]: Sorry, I was wrong, pushing the fix.

Comment (mathidachuk) [reply R/add_styles.R]: No it's my bad. I realized after I checked the rcmd that I never hit ctrl+d 🤦 🤦 🤦 

By the time I tried to push the changes you already fixed it. hehehhe thanks!

**Comment by tyluRp** [comment]
All tests have passed, merging! Thank you! 🎉
